modules/imgproc/include/opencv2/imgproc.hpp

```diff
@@ -3681,8 +3711,10 @@ data type.
 @param result Map of comparison results. It must be single-channel 32-bit floating-point. If image
 is \f$W \times H\f$ and templ is \f$w \times h\f$ , then result is \f$(W-w+1) \times (H-h+1)\f$ .
 @param method Parameter specifying the comparison method, see #TemplateMatchModes
-@param mask Mask of searched template. It must have the same datatype and size with templ. It is
-not set by default. Currently, only the #TM_SQDIFF and #TM_CCORR_NORMED methods are supported.
+@param mask Optional mask. It must have the same data type and size as templ. It must either have
+            the same number of channels as template or only one channel, which is then used for all
+            template and image channels. If the data type is #CV_8U, the mask values are divided by
+            255 . The exact formulas are documented in #TemplateMatchModes.
```

Comment:
For me it's very handy to just pass on the alpha channel of an image (which may be 8U) as mask, and therefore it makes sense to support the same type as the template, and also to divide by 255 for 8U (which makes no difference for the normalized modes). Interpreting 8U as binary mask and others not does not seem like a good idea to me either.

Also, looking at the implementation of `matchTemplate()` in its current form, I saw that all matrices (img, templ and mask) were divided by 255 in case of 8U. Should we keep that approach?

I also think a weight-mask instead of a simple binary mask adds much more flexibility (also some of this PR would not be neccessary as the implementation would be simpler). We could of course not call it "mask" but "weights" or "weight_mask" or something like that.

Any suggestions?

Reply:
I would suggest to split these behaviors by storage type:
- regular binary mask: CV_8U (values are 0 or 255). We should support that due OpenCV comparison functions.
- "weights" mask: CV_32F (assumed range 0..1)

P.S. Binary mask is converted to "weights" internally.

Reply:
Ok, I'll adapt the code, test and doc. I'm not really happy with it (behavior dependent on type) but I get your argument with compatibility with other 8U masks.

Reply:
I've shortened your proposed code to

        threshold(mask, mask, 0/*threshold*/, 1.0/*maxVal*/, THRESH_BINARY);
        mask.convertTo(mask, CV_32F);

without temp, and it seems to work fine. I hope there's no downside, otherwise let me know. I'll mark it as resolved.